src/lib/server/ring_buffer.hpp

```diff
@@ -0,0 +1,182 @@
+#pragma once
+
+#include <array>
+#include <boost/asio.hpp>  // NOLINT
+#include "postgres_message_types.hpp"
+
+namespace opossum {
+
+using Socket = boost::asio::ip::tcp::socket;
+
+static constexpr size_t BUFFER_SIZE = 4096u;
+
+// This class implements the logic of a circular buffer. If the end of the underlying data structure is reached
+// the iterator will start at the beginning if the data has already been processed.
+class RingBufferIterator : public std::iterator<std::forward_iterator_tag, char> {
+ public:
+  explicit RingBufferIterator(std::array<char, BUFFER_SIZE>& data, size_t position = 0)
+      : _data(data), _position(position) {}
+
+  RingBufferIterator& operator=(const RingBufferIterator& other) {
+    _position = other._position;
+    return *this;
+  }
+
+  RingBufferIterator(const RingBufferIterator& other) = default;
+
+  bool operator==(RingBufferIterator other) const { return (_position == other._position) && (_data == other._data); }
+
+  bool operator!=(RingBufferIterator other) const { return !(*this == other); }
+
+  RingBufferIterator& operator++() {
+    _position = (_position + 1) % BUFFER_SIZE;
+    return *this;
+  }
+
+  RingBufferIterator operator++(int) {
+    RingBufferIterator retval = *this;
+    ++(*this);
+    return retval;
+  }
+
+  RingBufferIterator& operator+(const size_t increment) {
+    _position = (_position + increment) % BUFFER_SIZE;
```

Comment:
This does not look correct. `operator+` should be const, which would correctly complain that the current iterator should not be modified. boost's iterator_facade makes it easier to correctly implement iterators